### PR TITLE
include: kernel, misc: Add queue API to uniquely append element

### PR DESCRIPTION
**Pull request by dhananjaygj**
Added the k_queue API to append unique element to the list. Existing k_queue_append API does not check if the element address already exists. This can create a problem if the same element address is appended to queue forming circular list giving unintended behavior for the application. The proposed API k_queue_unique_append takes care of checking if element exists already before appending. This API is complimentary to k_queue_remove which checks if the queue element is present before removing.

Signed-off-by: Dhananjay Gundapu Jayakrishnan dhananjay.jayakrishnan@proglove.de

### CODE REVIEW

File: include/kernel.h [hunk 1]
```diff
@@ -1977,6 +1977,30 @@ static inline bool k_queue_remove(struct k_queue *queue, void *data)
...
+ *
+ * @return true if data item was added, false if not
+ */
+static inline bool k_queue_unique_append(struct k_queue *queue, void *data)
+{
+	if (sys_sflist_is_node_in_list(&queue->data_q,
+			(sys_sfnode_t *)data) == false) {
+		k_queue_append(queue, data);
+		return true;
+	}
```
Comment (andyross) [inline include/kernel.h]: Can't you put the FOREACH search right here?  It wouldn't be meaningfully longer or more complicated.

**Review by andyross — CHANGES_REQUESTED** [review]
Not loving the mucking with slist.  We already have a "search this list" API that works fine.  I don't see much value in also having a "contains" test.

Mild -1.  If someone else really wants it in I'll remove it.  But it seems like k_queue_unique_append() should just have been written in terms of  SYS_SLIST_FOR_EACH_NODE

File: include/kernel.h (hunk 1 above)
Comment (dhananjaygj) [reply include/kernel.h]: Thanks for the review @andyross . Yes I could have directly implemented `k_queue_unique_append` using `SYS_SLIST_FOR_EACH_NODE`. I thought a "contains" test will be useful and hence implemented this way. But I also have the patch ready with implementation as you suggested here. I am waiting for @andrewboie's review on my PR

**Comment by dhananjaygj** [comment]
@andrewboie  Can you please take a look at my PR when you get some time?

**Comment by tbursztyka** [comment]
+1 on @andyross  comment, I don't see any other possible usage for node_in_list thingy so better not introducing it.

**Comment by dhananjaygj** [comment]
I updated the PR as suggested by @andyross and removed the node "node_in_list" test.

**Review by andyross — APPROVED** [review]
Thanks

**Pull request merged**